### ISSUE TITLE
Context manager usage

### ISSUE THREAD

**Issue by kkappler** [kujaku11/mth5#182]
TLDR: This following message regards an incorreect usage of the context manager:

When I attempt to use the context manager in the recommended way the process does not exit as desired.  here is a snippet:

```
    mth5_path = pathlib.Path.home().joinpath("software", "irismt","aurora", "data", "synthetic","mth5","test1.h5")
    with MTH5() as m:
        m = mth5.mth5.MTH5()
        m.open_mth5(mth5_path)
```
The output of this is:

> 2024-01-04T14:44:37.082257-0800 | WARNING | mth5.mth5 | filename | MTH5 file is not open or has not been created yet. Returning default name
2024-01-04T14:44:37.082402-0800 | INFO | mth5.mth5 | close_mth5 | Flushing and closing None
2024-01-04T14:44:37.087562-0800 | INFO | mth5.helpers | close_open_files | /home/kkappler/software/irismt/aurora/data/synthetic/mth5/test1.h5, Closed File

I followed the logic via the debugger, and it appears that by the time we have entered the `__exit__` method of MTH5,  the context manager _has already closed the file_.  

The call to `self.close_mth5()` within the `__exit__` method of MTH5 appears to be unneeded, and leads to a few confounding behaviours, (presumably due to the fact the context manager has closed the mth5):

1. m.filename property is not available, so we get a warning due to the line 
`self.logger.info(f"Flushing and closing {str(self.filename)}")` in `close_mth5()`
2. The line `self.__hdf5_obj.close()` in `close_mth5()` results in an `AttributeError` because `self.__hdf5_obj` is not available.  This in turn calls (via try/except) `helpers.close_open_files()`.  This is particularly concerning, because we maybe closing other h5s associated with other processes.

The solution _seems_ to be to remove the call to self.close_mth5()` within the `__exit__` method of MTH5.

I tested this, by running:
```
    mth5_path = pathlib.Path.home().joinpath("software", "irismt","aurora", "data", "synthetic","mth5","test1.h5")
    with MTH5() as m:
        m = mth5.mth5.MTH5()
        m.open_mth5(mth5_path)
    with MTH5() as m:
        m = mth5.mth5.MTH5()
        m.open_mth5(mth5_path)
    with MTH5() as m:
        m = mth5.mth5.MTH5()
        m.open_mth5(mth5_path)
```
and found no issues, i.e. opening the same h5 repeatedly did not cause any issues with opening and already open h5.
I also tested that the we could interact with the h5 repeatedly, i.e. I printed the channel summary a bunch of time.

However, this does not solve all issues ... when using the context manager, the close_mth5() loop is never entered, and this means that the channel summary is not built on close.  

I cannot see a truly robust way around this ... it would appear that the user needs to remember to either:

**A**: call `m.channel_summary.summarize()` if they were building their h5s using context manager 
OR
**B**: call `m.open_mth5()` and then `close_mth5()` after the build

Ideally, we would just open the mth5 and then close it again in the `__exit__` method, but knowledge of filename appears lost to the mth5 object by the time exit is encountered. 

In any case, the `self.close_mth5()` instruction in `__exit__` seems extraneous.

@kujaku11 : Might this be worked around by adding a `summarize_tables` method to the end of open_mth5()?

**Comment by kkappler** [comment]
A nice solution to this is actually to add `return self` to the end of mth5.MTH5.open_mth5()

This allows the following sort of syntax:

```
with MTH5().open_mth5(mth5_path) as m:
    do stuff
```
which I believe is preferable to how it was before:
```
with MTH5() as m:
    m = mth5.mth5.MTH5()
    m.open_mth5(args[0])
    do stuff
```

With this small change, the self.close_mth5 in `__exit__` executes correctly.

Will test and file a PR

**Comment by kkappler** [comment]
So, actually, the issue was that I was re-initializing the mth5 within the context, 

I should not have been doing:
```
mth5_path = pathlib.Path.home().joinpath("software", "irismt","aurora", "data", "synthetic","mth5","test1.h5")
    with MTH5() as m:
        m = mth5.mth5.MTH5()
        m.open_mth5(mth5_path)
    with MTH5() as m:
        m = mth5.mth5.MTH5()
        m.open_mth5(mth5_path)
    with MTH5() as m:
        m = mth5.mth5.MTH5()
        m.open_mth5(mth5_path)
```

But instead:

```
mth5_path = pathlib.Path.home().joinpath("software", "irismt","aurora", "data", "synthetic","mth5","test1.h5")
    with MTH5() as m:
        m.open_mth5(mth5_path)
    with MTH5() as m:
        m.open_mth5(mth5_path)
    with MTH5() as m:
        m.open_mth5(mth5_path)
```

**Comment by kujaku11** [comment]
Yes you should use the last snippet:

```
with MTH5(**kwargs) as m:
    m.open_mth5(file_path)
    Do stuff
```

**Comment by kkappler** [comment]
Yup :/

I'm ok with abandoning this PR, but I do like the one change that was made, which is to return self at the end of MTH5.open_mth5(), which allows for the other syntax as well:

```
with MTH5().open_mth5(mth5_path) as m:
    do stuff
```

**Comment by kkappler** [comment]
@kujaku11 Keeping context manger questions in the same issue, so reopening this one:

Do we know the consequences of opening an mth5 file of verison "0.1.0" as a "0.2.0"?

For example, if correct context manager usage is as follows:

```
with MTH5() as m:
    m.open_mth5(h5_path, mode=mode)
    do stuff
```

Then there is a possible issue since `MTH5()` takes an argument `file_version` which defaults to `0.2.0`.

Question: what if `h5_path` in the above was an existing file, that has file_version="0.1.0"?
Could this cause a problem?  Is file version of an existing archive mutable?

Some tests that address this should be added.
It would be nice if the file already exists that

**Comment by kujaku11** [comment]
@kkappler This is a good question, I think that the tests pass because its just the metadata key word `file_version` that is changing. I'm not sure there are any internal checks on the file structure when `file_version` is changed.  

For example if you have an existing v1 file and want to convert it to a v2 just by changing the `file_version` I don't think that simply adds a `Surveys` group and moves the stations into that survey.  I think that you would need to create a whole new file.  Which brings us back to the external link option where if you have an v1 file and you want to make it a v2 you could create a new v2 file with an external link to the v1 file under the correct location `Experiment/Surveys/survey_name_v1/station_v1`. Going from v2 to v1 I think you need to create a whole new file.

Will have to add some checks when the `file_version` is changed.  I think we should make the file version unmutable for now otherwise it will cause confusion.
